### PR TITLE
chore: kubernetes_logs rewrite RFC

### LINKED ISSUE #10298: kubernetes_logs source rewrite RFC

### PR DESCRIPTION
**Pull request by spencergilbert**
Closes #10298

### CODE REVIEW

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+### Replace existing source with an enrichment transform
+
+This option is capable of being used regardless of what role Vector has been deployed
+in, but it is a larger change to implement and a larger update for users too. Longer
+term the source can be decoupled and introduced as a standalone enrichment component.
+
+## Outstanding Questions
+
+- Do we want to make this change with a version change of the source, opting in
+to the change with a `version` option in the configuration?
```
Comment (JeanMertz) [inline rfcs/2021-12-06-10298-kubernetes_logs-rewrite:102]: If our tests pass, our soaks show no regression, and there are nu user-facing changes, then I see no reason why we'd want to 1) keep the old code around, and 2) have users be confused by this knob.

Or is there another reason you're thinking of why we might want to do this?

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+## Plan Of Attack
+
+- [ ] ...
+
+## Future Improvements
+
+- Move Kubernetes enrichment to a standalone transform and rework source to be
+a more simple combination of file source plus the new enrichment transform
```
Comment (JeanMertz) [inline rfcs/2021-12-06-10298-kubernetes_logs-rewrite:111]: Would work on this migration be _easier_ if we did this step first?

**Review by JeanMertz — APPROVED** [review]
I'm for this change, but the big caveat is that it'll be a business decision in the end whether the required investment is worth the trade-off, as we won't gain any immediate benefit from this change.

That is to say, I think we should merge this RFC with the intent for this work to happen at some point, but there are many other things that I suspect will take priority for the next ~2 quarters.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite (hunk 1 above)
Comment (spencergilbert) [reply rfcs/2021-12-06-10298-kubernetes_logs-rewrite:102]: Only as a more cautious style of transition, we could handle this internally by adding a new feature for the rewrite. I expect it's an overly cautious approach to expose it as a new "version" of the component 👍

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite (hunk 2 above)
Comment (spencergilbert) [reply rfcs/2021-12-06-10298-kubernetes_logs-rewrite:111]: That's a good question - architecturally it a greater shift (obviously), and we'd also need to address the transition for users. Would we continue to have a source that was internally a preconfigured `file` source followed immediately by the new transform? Would we just guide users on a reasonable configuration for the `file` source to replicate the old behavior?

The enrichment transform seems a better pattern for aggregators but less so agents, so similar to your other comment - seems like there's some business decision to be make here as well.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite (hunk 1 above)
Comment (jszwedko) [reply rfcs/2021-12-06-10298-kubernetes_logs-rewrite:102]: Yeah, I can see it as a derisking approach. I personally think the risk here is low enough that I don't think we need to keep the old code around and version it. This is in contrast to the disk buffer rewrite where it is risky enough that we want to keep both around for a little while at least.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite (hunk 2 above)
Comment (jszwedko) [reply rfcs/2021-12-06-10298-kubernetes_logs-rewrite:111]: I'd rather just focus on re-implementing the `kubernetes_logs` source as-is first and then split out later if we find the motivation too. I think this is less risky and doesn't require additional design thought.

**Comment by jszwedko** [comment]
Looking good! To Jean's point about prioritization, it also depends on the level of effort we think this is. Do you have a rough idea of how long it'd take to do the re-implementation?

**Comment by spencergilbert** [comment]
> Looking good! To Jean's point about prioritization, it also depends on the level of effort we think this is. Do you have a rough idea of how long it'd take to do the re-implementation?

I'd hope no more than a sprint, two weeks?

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite (hunk 1 above)
Comment (jszwedko) [reply rfcs/2021-12-06-10298-kubernetes_logs-rewrite:102]: I'd add the answer to the open question here.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+term the source can be decoupled and introduced as a standalone enrichment component.
+
+## Outstanding Questions
+
+- Do we want to make this change with a version change of the source, opting in
+to the change with a `version` option in the configuration?
+
+## Plan Of Attack
+
+- [ ] ...
```
Comment (jszwedko) [inline rfcs/2021-12-06-10298-kubernetes_logs-rewrite:106]: Reminder to fill this in.

**Review by jszwedko — APPROVED** [review]
I left a couple of comments but approving since this is pretty much there.

**Comment by blt** [comment]
Maybe I'm taking crazy pills here but is the RFC file missing its `.md` extension?

**Comment by spencergilbert** [comment]
> Maybe I'm taking crazy pills here but is the RFC file missing its `.md` extension?

🙈 _no idea what you're talking about_

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+### Out of scope
+
+- Changing the end-to-end functionality of the `kubernetes_logs` source
+- Changing the testing strategies we have around the `kubernetes_logs` source today
+
+## Pain
+
+- Difficulty troubleshooting and supporting this feature as maintainers
+- Unresolved bug reports from users related to our in-house Kubernetes code
```
Comment (binarylogic) [inline rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md:24]: Can you list out the specific bug reports? That'll help gauge the value of this change.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+enough to not warrant a version split between the old and new code.
+
+## Plan Of Attack
+
+- Replace contents of `src/kubernetes` with equivalents from `kube`
+- Ensure unit tests and integration tests show matching behavior before and after rewrite
+
+## Future Improvements
+
+- Move Kubernetes enrichment to a standalone transform and rework source to be
```
Comment (binarylogic) [inline rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md:112]: Let's also break this out into incremental steps. I'm curious what it looks like.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+## Outstanding Questions
+
+- ~~Do we want to make this change with a version change of the source, opting in
+to the change with a `version` option in the configuration?~~ No, the risk seems low
+enough to not warrant a version split between the old and new code.
+
+## Plan Of Attack
+
+- Replace contents of `src/kubernetes` with equivalents from `kube`
```
Comment (binarylogic) [inline rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md:107]: Let's also break this out into incremental steps. I'm curious what it looks like.

Comment (spencergilbert) [reply rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md:107]: I'll be taking another pass and also responding to #2963 reasons for _not_ using `kube`.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+### Replace existing source with an enrichment transform
+
+This option is capable of being used regardless of what role Vector has been deployed
+in, but it is a larger change to implement and a larger update for users too. Longer
+term the source can be decoupled and introduced as a standalone enrichment component.
+
+## Outstanding Questions
+
+- Do we want to re-evaluate the usage of `evmap` in this component?
+- ~~Do we want to make this change with a version change of the source, opting in
```
Comment (binarylogic) [inline rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md:120]: Agree, if the change is internal then it should not break the public API.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+### Replace existing source with an enrichment transform
+
+This option is capable of being used regardless of what role Vector has been deployed
+in, but it is a larger change to implement and a larger update for users too. Longer
+term the source can be decoupled and introduced as a standalone enrichment component.
+
+## Outstanding Questions
+
+- Do we want to re-evaluate the usage of `evmap` in this component?
```
Comment (binarylogic) [inline rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md:119]: Do you have an answer for this?

**Review by binarylogic — APPROVED** [review]
Looks good to me.

File: rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md (hunk 8 above)
Comment (spencergilbert) [reply rfcs/2021-12-06-10298-kubernetes_logs-rewrite.md:119]: I _don't_ think I do, I don't remember seeing any of the reasoning why we picked `evmap` originally. It seems like it's possibly higher performance depending on usage. We could probably benchmark further down the line, `kubernetes_logs` is the only thing in Vector that uses it, so it'd mostly be reducing dependencies.

**TL;DR:** probably fine keeping it as is, unless we strongly want to remove deps that may be superfluous.
